### PR TITLE
tracing examples with OTEL Splunk java agent

### PR DESCRIPTION
**Pull request by kubawach**
Resolves APMI-843

Same examples as for SignalFX Agent, but with OTEL-specific code (manual), configurations and easy-to-use startup scripts.

### CODE REVIEW

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+$ export OTEL_ZIPKIN_ENDPOINT=http://localhost:9080/v1/trace
+
+$ # Replace this with your application's common name
+$ export OTEL_ZIPKIN_SERVICE_NAME=unnamed-java-app
+```
+
+To obtain the latest Java agent, download it to your host's filesystem:
+
+```
+curl -sSL -o /opt/opentelemetry-javaagent-all.jar 'https://github.com/signalfx/signalfx-otel-java/releases/latest/download/signalfx-otel-javaagent-all.jar'
```
Comment (iNikem) [inline opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:25]: This will not until repo is made public

File: opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/src/main/kotlin/com/splunk/tracing/otel/examples/wishlist/Controllers.kt [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            response.status = HttpServletResponse.SC_NOT_FOUND
+        }
+
+        model["wishers"] = wishers.sortedWith(compareBy{it.lastName})
+        return "wishlist"
+    }
+
+    @PostMapping("/wishers/delete/{wisherId}")
+    fun deleteById(@PathVariable wisherId: Long, response: HttpServletResponse, model: Model): String {
+        var wisher : Wisher
```
Comment (mateuszrzeszutek) [inline opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/src/main/kotlin/com/splunk/tracing/otel/examples/wishlist/Controllers.kt:82]: Can be `val` inside try

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                .url(url)
+                .build();
+
+        // This span acts as a root span that envelops the sets of spans generated
+        // by both the HTTP call (using the OKHttp auto-instrumentation) and the
+        // Redis set action (using the redis-jedis auto-instrumentation).  If you
+        // did not have this root span that is manaully created, you would instead
+        // see two independent traces for the HTTP get and the Redis set.
+        // The Java agent sets the GlobalTracer instance before the application's main method is called.
+        Span span = OpenTelemetry.getTracer("sample").spanBuilder("fetch-and-set").startSpan();
```
Comment (iNikem) [inline opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java:65]: Store `OpenTelemetry.getTracer("sample")` to a local variable. It is bad example to call it twice in a row

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+
+        try {
+            executor.shutdown();
+            executor.awaitTermination(3000, TimeUnit.MILLISECONDS);
+            redisClient.close();
+            // Give the tracer time to flush the spans in the sender thread.  In
+            // long-running apps this is generally unnecessary but unfortunately there is no
+            // OpenTracing interface method for closing/stopping a tracer.
```
Comment (iNikem) [inline opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java:91]: `OpenTracing`?

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/pom.xml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    <dependency>
+      <groupId>io.opentelemetry</groupId>
+      <artifactId>opentelemetry-api</artifactId>
+      <version>0.8.0</version>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.opentelemetry</groupId>
+      <artifactId>opentelemetry-sdk</artifactId>
```
Comment (iNikem) [inline opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/pom.xml:34]: I don't know if this is important here or not, but technically you don't need SDK if you plan to run your application with javaagent anyway. API should be enough.

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+```sh
+$ # Export the environment variables in this shell as shown above for your org
+$ mvn package
+
+$ # Run a Redis server locally that the example app will use
+$ docker run -d --name redis-tracing-test -p 6379:6379 redis
+
+$ java -javaagent:/opt/opentelemetry-javaagent-all.jar -jar target/java-agent-example-1.0-SNAPSHOT-shaded.jar https://google.com
+```
```
Comment (iNikem) [inline opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:44]: Shouldn't we tell user to execute `start.sh` instead?

File: opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/pom.xml [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    </parent>
+    <groupId>com.splunk.tracing.otel.examples</groupId>
+    <artifactId>kotlin-spring-demo</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+    <name>Wishlist</name>
+    <description>Example Spring Boot application in Kotlin</description>
+
+    <properties>
+        <java.version>1.8</java.version>
+        <kotlin.version>1.2.71</kotlin.version>
```
Comment (iNikem) [inline opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/pom.xml:19]: That is ancient kotlin version, can we upgrade?

File: opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/postgres/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+version: '2'
```
Comment (iNikem) [inline opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/postgres/docker-compose.yml:1]: Why do we need docker-compose for 1 service?

File: signalfx-tracing/signalfx-java-tracing/jetty-customtags/run-server.sh [hunk 9]
```diff
@@ -6,4 +6,4 @@ export SIGNALFX_SERVICE_NAME=EchoServer-JavaTagDemo
 
 mvn compile exec:exec \
   -Dexec.executable="java" \
-  -Dexec.args="-javaagent:/opt/signalfx-tracing.jar -cp %classpath sf.main.EchoDemo 5000"
+  -Dexec.args="-javaagent:/opt/signalfx-tracing.jar -Dsignalfx.endpoint.url=http://127.0.0.1:8888 -cp %classpath sf.main.EchoDemo 5000"
```
Comment (mateuszrzeszutek) [inline signalfx-tracing/signalfx-java-tracing/jetty-customtags/run-server.sh:9]: Why this change?

File: opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/start.sh [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#!/bin/zsh
+
+export OTEL_EXPORTER=zipkin
+export OTEL_ZIPKIN_ENDPOINT=http://localhost:8888
```
Comment (iNikem) [inline opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/start.sh:4]: This is untraditional port for SA and contradicts README

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        Request request = new Request.Builder()
+                .url(url)
+                .build();
+
+        // This span acts as a root span that envelops the sets of spans generated
+        // by both the HTTP call (using the OKHttp auto-instrumentation) and the
+        // Redis set action (using the redis-jedis auto-instrumentation).  If you
+        // did not have this root span that is manaully created, you would instead
+        // see two independent traces for the HTTP get and the Redis set.
+        // The Java agent sets the GlobalTracer instance before the application's main method is called.
```
[the remark this answers is not in the document]
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java:64]: Great catch.

File: opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/src/main/kotlin/com/splunk/tracing/otel/examples/wishlist/Controllers.kt [hunk 12]
```diff
@@ -0,0 +1,139 @@
+package com.splunk.tracing.otel.examples.wishlist
+
+import org.springframework.stereotype.Controller
+import org.springframework.validation.BindingResult
+import org.springframework.ui.Model
+import org.springframework.ui.set
+import org.springframework.web.bind.annotation.*
```
[the remark this answers is not in the document]
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/src/main/kotlin/com/splunk/tracing/otel/examples/wishlist/Controllers.kt:7]: Out of scope - this is direct copy of existing SFx examples.

File: opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/src/main/kotlin/com/splunk/tracing/otel/examples/wishlist/Controllers.kt (hunk 2 above)
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/src/main/kotlin/com/splunk/tracing/otel/examples/wishlist/Controllers.kt:82]: Out of scope - this is a direct copy of existing SFx examples.

File: signalfx-tracing/signalfx-java-tracing/jetty-customtags/run-server.sh (hunk 9 above)
Comment (kubawach) [reply signalfx-tracing/signalfx-java-tracing/jetty-customtags/run-server.sh:9]: Git is stupid, that's why (will revert ofc ;) ).

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+modification of the target application's source code.
+
+Here are some environment variables you might need to set to configure the Java
+agent:
+```
+$ # This is the ZIPKIN exporter - compatible with SignalFX SmartAgent
+$ export OTEL_EXPORTER=zipkin
+
+$ # This is the default endpoint url
+$ export OTEL_ZIPKIN_ENDPOINT=http://localhost:9080/v1/trace
```
[the remark this answers is not in the document]
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:16]: I'll change the wording to that it's clear that user can set these values if needed.

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md (hunk 1 above)
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:25]: Yes, I understand we want to release these examples along with our distro of the OTEL agent (correct me if I'm wrong!).

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md (hunk 6 above)
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:44]: Will add note.

File: opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/postgres/docker-compose.yml (hunk 8 above)
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/postgres/docker-compose.yml:1]: It's taken directly from the SFx example.

File: opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/start.sh (hunk 10 above)
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/start.sh:4]: Reverted to RM value.

File: opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/pom.xml (hunk 7 above)
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/spring-boot-kotlin/pom.xml:19]: Set the newest one.

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java:91]: I spoke with Nikita and we could avoid the wait by either calling forceFlush() or adding a span processor. However in both cases the app would need to use SDK directly which is not recommended for an app (should make use of API only). Therefore at this moment I believe it's the only way.

**Comment by kubawach** [comment]
FYI @mateuszrzeszutek @iNikem updated the PR with code review changes.

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md [hunk 14]
```diff
@@ -9,14 +9,17 @@ modification of the target application's source code.
 Here are some environment variables you might need to set to configure the Java
 agent:
 ```
-$ # This is the ZIPKIN exporter - compatible with SignalFX SmartAgent
+$ # Exporter configuration - default is OTEL exporter
```
Comment (iNikem) [inline opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:12]: OTLP exporter

**Review by iNikem — APPROVED** [review]
These examples don't provide any destination for telemetry data, right? Maybe we want to include running Otel Collector into examples? E.g. if we already have docker compose file, it may run Collector as well. WDYT?

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md (hunk 14 above)
Comment (mateuszrzeszutek) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:12]: Also, by default Splunk distribution uses Zipkin exporter: https://github.com/signalfx/splunk-otel-java/blob/main/agent/src/main/java/com/splunk/opentelemetry/SplunkAgent.java#L31

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+Here are some environment variables you might need to set to configure the Java
+agent:
+```
+$ # Exporter configuration - default is OTEL exporter
+$ # Eg. set ZIPKIN exporter - compatible with SignalFX SmartAgent
+$ export OTEL_EXPORTER=zipkin
+
+$ # Exporter endpoint URL - each exporter has own, specific property name and the default value  
+$ # Eg. set ZIPKIN exporter URL - default is http://localhost:9411/api/v2/spans
```
Comment (mateuszrzeszutek) [inline opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:17]: This is no longer true - see https://github.com/signalfx/splunk-otel-java/blob/main/agent/src/main/java/com/splunk/opentelemetry/SplunkAgent.java#L34

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+
+        Tracer tracer = OpenTelemetry.getTracer("sample");
+
+        // This span acts as a root span that envelops the sets of spans generated
+        // by both the HTTP call (using the OKHttp auto-instrumentation) and the
+        // Redis set action (using the redis-jedis auto-instrumentation).  If you
+        // did not have this root span that is manually created, you would instead
+        // see two independent traces for the HTTP get and the Redis set.
+        // The Java agent sets the Tracer instance before the application's main method is called.
```
Comment (mateuszrzeszutek) [inline opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java:68]: I'd just remove this sentence -- it does not provide any useful information.

Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java:68]: Wait, what? Do you assume that each newbie will know all (manual paren span, how does Tracer instance end up in runtime) that? I'm not sure I agree :)

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md (hunk 15 above)
Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:17]: It's hard do judge if we should refer to general agent or Splunk (temporary?) distribution. I decided to stick with general behaviour, but can change that of course. @iNikem what do you think?

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java (hunk 16 above)
Comment (mateuszrzeszutek) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/src/main/java/com/splunk/tracing/otel/examples/javaagent/App.java:68]: I meant this 
```
// The Java agent sets the Tracer instance before the application's main method is called.
```
sentence: the Java agent does not actually set it in any way, it's not a global variable (like `GlobalTracer` was), you're just using OTel API directly.

File: opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md (hunk 15 above)
Comment (iNikem) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:17]: If this example references our (Splunk) distribution, then it should be consistent with its behaviour. Including defaults.

Comment (kubawach) [reply opentelemetry-tracing/opentelemetry-java-tracing/okhttp-and-jedis/README.md:17]: Changed to "splunk default" values.

**Comment by kubawach** [comment]
Task completed - changed the README once again and adjusted existing startup scripts as well.
